phpunit/FuzzyTestSuiteLoader.php

```diff
@@ -0,0 +1,146 @@
+<?php
+/**
+ * This file is derived from the PHPUnit project and customized for the
+ * needs of the Gutenberg project.
+ *
+ * @package gutenberg
+ */
+
+namespace phpunit;
+
+use PHPUnit\Framework\TestCase;
+use PHPUnit\Util\FileLoader;
+use PHPUnit\Runner\TestSuiteLoader;
+
+use \Exception;
+use \ReflectionClass;
+use \ReflectionException;
+
+/*
+ * Let's disable the conflicting Gutenberg phpcs inspections.
+ *
+ * The code in this file conforms to PHPUnit coding standards.
+ * We need to preserve the original signatures with type annotations
+ * and the camel-cased arguments names, as renaming parameters is
+ * considered a breaking change in PHP8.
+ */
+// phpcs:disable PHPCompatibility.FunctionDeclarations.NewParamTypeDeclarations.stringFound
+// phpcs:disable PHPCompatibility.FunctionDeclarations.NewReturnTypeDeclarations.stringFound
+// phpcs:disable PHPCompatibility.FunctionDeclarations.NewReturnTypeDeclarations.class_nameFound
+// phpcs:disable WordPress.NamingConventions.ValidVariableName.InterpolatedVariableNotSnakeCase
+// phpcs:disable WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
+
+/**
+ * @internal This class is not covered by the backward compatibility promise for PHPUnit
+ *
+ * @deprecated see https://github.com/sebastianbergmann/phpunit/issues/4039
+ */
+class FuzzyTestSuiteLoader implements TestSuiteLoader {
+
+	/**
+	 * @throws Exception When no test class is detected in the $suite_class_file.
+	 */
+	public function load( string $suiteClassFile ): ReflectionClass {
+		$suiteClassName = $this->findTestClass( $suiteClassFile );
+
+		try {
+			$class = new ReflectionClass( $suiteClassName );
+		} catch ( ReflectionException $e ) {
+			throw new Exception(
+				$e->getMessage(),
+				(int) $e->getCode(),
+				$e
+			);
+		}
+
+		if ( $class->isSubclassOf( TestCase::class ) && ! $class->isAbstract() ) {
+			return $class;
+		}
+
+		if ( ! $class->hasMethod( 'suite' ) ) {
+			throw new Exception(
+				"Class '$suiteClassName' could not be found in '$suiteClassFile'."
+			);
+		}
+
+		try {
+			$method = $class->getMethod( 'suite' );
+		} catch ( ReflectionException $e ) {
+			throw new Exception(
+				$e->getMessage(),
+				(int) $e->getCode(),
+				$e
+			);
+		}
+
+		if ( $method->isAbstract() || ! $method->isPublic() || ! $method->isStatic() ) {
+			throw new Exception(
+				"Class '$suiteClassName' could not be found in '$suiteClassFile'."
+			);
+		}
```

Comment:
do you think we could reuse the upstream version of this function after transforming the filename?

```php
public function load( string $suiteClassFile ): ReflectionClass {
	return parent::load( $this->normalizeSuiteClassFile );
}

protected function normalizeSuiteClassFile( $suiteClassFile ): string 
	$suiteClassFile = str_replace( '-', '_', $suiteClassFile );
	$suiteClassFile = preg_replace_callback(
		'/^\w|_\w/',
		function ( $matches ) { return strtoupper( $matches[0] ); }
		$suiteClassFile
	);
}
```

I only ask this because it doesn't seem like PHPUnit needs to load the file and if we "trick" it like this we don't have to re-implement `load`

Reply:
well if we can do this it doesn't look easy because the `TestSuiteLoader` is a `final` class we can't extend. I tried doing so and failed a few different ways. oh well.

Reply:
`final` seems like a weird choice – perhaps we could submit a PR to make it non-final eventually? It wouldn't solve the problem at hand but could be worth it in the future.